private/bufpkg/bufcas/digest.go

```diff
@@ -0,0 +1,240 @@
+		shakeHash := sha3.NewShake256()
+		shakeHash.Reset()
+		if _, err := io.Copy(shakeHash, reader); err != nil {
+			return nil, err
+		}
+		value := make([]byte, shake256Length)
+		if _, err := shakeHash.Read(value); err != nil {
+			// sha3.ShakeHash never errors or short reads. Something horribly wrong
+			// happened if your computer ended up here.
+			return nil, err
+		}
+		return newDigest(digestType, value)
+	default:
+		return nil, fmt.Errorf("unknown DigestType: %v", digestType)
+	}
+}
+
+// NewDigestForString returns a new Digest for the given Digest string.
+//
+// The string is expected to be non-empty, If not, an error is treutned.
+//
+// This reverses Digest.String().
+// A Digest string is of the form typeString:hexValue.
+func NewDigestForString(s string) (Digest, error) {
+	if s == "" {
+		return nil, errors.New("empty string passed to NewDigestForString")
+	}
+	digestTypeString, hexValue, ok := strings.Cut(s, ":")
+	if !ok {
+		return nil, fmt.Errorf("invalid Digest string: %q", s)
+	}
+	digestType, err := ParseDigestType(digestTypeString)
+	if err != nil {
+		return nil, err
+	}
+	value, err := hex.DecodeString(hexValue)
+	if err != nil {
+		return nil, err
+	}
+	return newDigest(digestType, value)
+}
+
+// DigestToProto converts the given Digest to a proto Digest.
+//
+// TODO: validate the returned Digest.
+func DigestToProto(digest Digest) (*storagev1beta1.Digest, error) {
+	protoDigestType, ok := digestTypeToProto[digest.Type()]
+	// Technically we have aleady done this validation but just to be safe.
+	if !ok {
+		return nil, fmt.Errorf("unknown DigestType: %v", digest.Type())
+	}
+	return &storagev1beta1.Digest{
+		Type:  protoDigestType,
+		Value: digest.Value(),
+	}, nil
+}
+
+// ProtoToDigest converts the given proto Digest to a Digest.
+//
+// Validation is performed to ensure the DigestType is known, and the value
+// is a valid digest value for the given DigestType.
+// TODO: validate the input proto Digest.
+func ProtoToDigest(protoDigest *storagev1beta1.Digest) (Digest, error) {
+	digestType, ok := protoToDigestType[protoDigest.Type]
+	if !ok {
+		return nil, fmt.Errorf("unknown proto Digest.Type: %v", protoDigest.Type)
+	}
+	return NewDigest(digestType, protoDigest.Value)
+}
+
+// DigestEqual returns true if the given Digests are considered equal.
+//
+// If both Digests are nil, this returns true.
+//
+// This check both the DigestType and Digest value.
+func DigestEqual(a Digest, b Digest) bool {
+	if (a == nil) != (b == nil) {
+		return false
+	}
+	if a == nil {
+		return true
+	}
+	if a.Type() != b.Type() {
+		return false
+	}
+	return bytes.Equal(a.Value(), b.Value())
+}
+
+/// *** PRIVATE ***
+
+type digest struct {
+	digestType DigestType
+	value      []byte
+	// Cache as we call String pretty often.
+	// We could do this lazily but not worth it.
+	stringValue string
+}
+
+func newDigest(digestType DigestType, value []byte) (*digest, error) {
+	switch digestType {
+	case DigestTypeShake256:
```

Comment:
I would argue otherwise in this case: generally I think it's good to set up `switches` for enums, as its a clear indication to future developers that this is a forked codepath based on the value of the enum. An `if` statement isn't as clear, and in my opinion having an `if` statement in these cases leads to future bugs when new enum values are added (the developer does not realize the `if` statement needs adjusting).